tests/unit/store/http_client/test_candid_client.py

```diff
@@ -0,0 +1,258 @@
+    CandidConfig,
+    _http_client,
+)
+
+
+def test_config_section_name():
+    assert CandidConfig()._get_section_name() == "dashboard.snapcraft.io"
+
+
+def test_config_section_name_with_env(monkeypatch):
+    monkeypatch.setenv("STORE_DASHBOARD_URL", "http://dashboard.other.com")
+
+    assert CandidConfig()._get_section_name() == "dashboard.other.com"
+
+
+def test_config_path(xdg_dirs):
+    assert (
+        CandidConfig()._get_config_path() == xdg_dirs / ".config/snapcraft/candid.cfg"
+    )
+
+
+def test_candid_client_has_no_credentials(xdg_dirs):
+    assert CandidClient().has_credentials() is False
+
+
+def test_candid_client_has_credentials(xdg_dirs):
+    # Baseline check.
+    assert CandidClient().has_credentials() is False
+
+    # Setup.
+    client = CandidClient()
+    client._macaroon = "macaroon"
+    client._auth = "auth"
+
+    assert client.has_credentials() is True
+    assert CandidClient().has_credentials() is True
+
+
+@pytest.fixture
+def candid_client(xdg_dirs, monkeypatch):
+    """Return a CandidClient with an alterate requests method."""
+    bakery_client = Mock(spec=httpbakery.Client)
+
+    def mock_discharge(*args, **kwargs):
+        return [
+            Macaroon(
+                location="api.snapcraft.io",
+                signature="d9533461d7835e4851c7e3b639144406cf768597dea6e133232fbd2385a5c050",
+            )
+        ]
+
+    monkeypatch.setattr(bakery, "discharge_all", mock_discharge)
+
+    return CandidClient(bakery_client=bakery_client)
+
+
+@pytest.fixture
+def snapcraft_macaroon():
+    return json.dumps(
+        {
+            "s64": "a0Vi7CwhHWjS4bxzKPhCZQIEJDvlbh9FyhOtWx0tNFQ",
+            "c": [
+                {"i": "time-before 2022-03-18T19:54:57.151721Z"},
+                {
+                    "v64": "pDqaL9KDrPfCQCLDUdPc8yO2bTQheWGsM1tpxRaS_4BT3r6zpdnT5TelXz8vpjb4iUhTnc60-x5DPKJOpRuwAi4qMdNa67Vo",
+                    "l": "https://api.jujucharms.com/identity/",
+                    "i64": "AoZh2j7mbDQgh3oK3qMqoXKKFAnJvmOKwmDCNYHIxHqQnFLJZJUBpqoiJtqra-tyXPPMUTmfuXMgOWP7xKwTD26FBgtJBdh1mE1wt3kf0Ur_TnOzbAWQCHKxqK9jAp1jYv-LlLLAlQAmoqvz9fBf2--dIxHiLIRTThmAESAnlLZHOJ7praDmIScsLQC475a85avA",
+                },
+                {
+                    "i": 'extra {"package_id": null, "channel": null, "acl": ["package_access", "package_manage", "package_push", "package_register", "package_release", "package_update"], "store_ids": null}'
+                },
+            ],
+            "l": "api.snapcraft.io",
+            "i64": "AwoQ2Ft5YBjnovqdr8VNV3TSlhIBMBoOCgVsb2dpbhIFbG9naW4",
+        }
+    )
+
+
+def test_login_discharge_macaroon(candid_client, snapcraft_macaroon):
+    candid_client.request
+    candid_client.login(macaroon=snapcraft_macaroon)
+
+    assert candid_client.has_credentials() is True
+    assert candid_client._macaroon == snapcraft_macaroon
+    assert candid_client._auth == (
+        "W3siaWRlbnRpZmllciI6ICIiLCAic2lnbmF0dXJlIjogImQ5NTMzNDYxZDc4MzVlNDg1MWM"
```

Comment:
Can this be done in a way to show how this magic string is derived? It's not obvious when looking at the tests.

Reply:
I can add a code comment, it's not magic